src/rpc/rpc.cc

```diff
@@ -7,25 +7,150 @@
 
 #include <dgl/runtime/container.h>
 #include <dgl/packed_func_ext.h>
+#include <dgl/zerocopy_serializer.h>
 #include "../c_api_common.h"
 
+using dgl::network::StringPrintf;
 using namespace dgl::runtime;
 
 namespace dgl {
 namespace rpc {
 
 RPCStatus SendRPCMessage(const RPCMessage& msg) {
-  // TODO(???): Implement
+  std::string zerocopy_blob;
+  StringStreamWithBuffer zc_write_strm(&zerocopy_blob);
+  static_cast<dmlc::Stream *>(&zc_write_strm)->Write(msg);
+  char* rpc_meta_buffer = new char[zerocopy_blob.size()];
+  // Copy the data for now, can be optimized later
+  memcpy(rpc_meta_buffer, zerocopy_blob.data(), zerocopy_blob.size());
+  int64_t rpc_meta_size = zerocopy_blob.size();
+  network::Message rpc_meta_msg;
+  rpc_meta_msg.data = rpc_meta_buffer;
+  rpc_meta_msg.size = rpc_meta_size;
+  rpc_meta_msg.deallocator = network::DefaultMessageDeleter;
+  CHECK_EQ(RPCContext::ThreadLocal()->sender->Send(
+    rpc_meta_msg, msg.server_id), ADD_SUCCESS);
+  // send ndarray count
+  char* ndarray_count = new char[sizeof(int32_t)];
+  *(reinterpret_cast<int32_t*>(ndarray_count)) = msg.tensors.size();
+  network::Message ndarray_count_msg;
+  ndarray_count_msg.data = ndarray_count;
+  ndarray_count_msg.size = sizeof(int32_t);
+  ndarray_count_msg.deallocator = network::DefaultMessageDeleter;
+  CHECK_EQ(RPCContext::ThreadLocal()->sender->Send(
+    ndarray_count_msg, msg.server_id), ADD_SUCCESS);
+  // send real ndarray data
+  for (auto ptr : zc_write_strm.buffer_list()) {
+    network::Message ndarray_data_msg;
+    ndarray_data_msg.data = reinterpret_cast<char*>(ptr.data);
+    ndarray_data_msg.size = ptr.size;
+    NDArray tensor = ptr.tensor;
+    ndarray_data_msg.deallocator = [tensor](network::Message*) {};
+    CHECK_EQ(RPCContext::ThreadLocal()->sender->Send(
+      ndarray_data_msg, msg.server_id), ADD_SUCCESS);
+  }
   return kRPCSuccess;
 }
 
 RPCStatus RecvRPCMessage(RPCMessage* msg, int32_t timeout) {
-  // TODO(???): Implement
+  // ignore timeout now
+  network::Message rpc_meta_msg;
+  int send_id;
+  CHECK_EQ(RPCContext::ThreadLocal()->receiver->Recv(
+    &rpc_meta_msg, &send_id), REMOVE_SUCCESS);
+  // Copy the data for now, can be optimized later
```

Comment:
`// TODO(chao): ...`  I guess this requires @VoVAllen 's modification on the stream class right?

Reply:
Yes